### PR TITLE
Add support for grub2-mkconfig on Mariner

### PR DESCRIPTION
**Pull request by AZaugg**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [X] The toolchain has been rebuilt successfully (or no changes were made to it)
- [X] The toolchain/worker package manifests are up-to-date
- [X] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [X] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [X] All package sources are available
- [X] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [X] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [X] All source files have up-to-date hashes in the `*.signatures.json` files
- [X] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [X] Documentation has been updated to match any changes to the build system
- [X] Ready to merge

---

###### Summary 

Today, when managing grub config files via the official grub toolchain, (grub2-mkconfig), it updates and replaces the existing  /boot/grub/grub.cfg that is laid out by the initial installer, in turn if a user was to use the official tool that is expected to work on other distros on Mariner, the user would be left with an unbootable system. Essentially the official toolchain that people have come to expect is not compatible withs Mariner.  This commit brings support for the industry default tool chain used to manage grub configs.

###### Change Log  

- Made grub tools grub2-mkconfig workable with Mariner grub install

###### Does this affect the toolchain?  

**NO**

###### Associated issues  

- #xxxx

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-YYYY-XXXX

###### Test Methodology

- Build Grub2 package on a mariner system with my changes
- confirmed flle /etc/default/grub was written out with defaults
- ran grub2-mkconfig
- rebooted system
- System came up fine

### CODE REVIEW

**Comment by AZaugg** [comment]
Not actually sure why the spec lint fails. It runs fine on my build machine

File: SPECS/grub2/grub2.spec [hunk 1]
```diff
@@ -269,9 +269,31 @@ GRUB_PXE_MODULE_SOURCE=%{buildroot}%{_datadir}/grub2-efi/grubaa64-noprefix.efi
...
 
-%post   -p /sbin/ldconfig
+%post
+/sbin/ldconfig
+# When grub2 is installed, we need to populate the defaults file, assuming that a user
+# may invoke grub tools such as grub2-mkconfig for grub updates, in its current state, running
+# grub2-mkconfig on Mariner will brick Mariner, as there is an assumption that the grub.cfg file
+# is setup and configured only once during bootstrap of the image.
+
+# Only update the file if the file exists and is empty, take last known good CMDLINE configuration
```
Comment (ddstreet) [inline SPECS/grub2/grub2.spec:279]: The way to do this is to list the file in the normal %files section, but as a config file.
https://rpm-software-management.github.io/rpm/manual/spec.html#config

File: SPECS/grub2/grub2.spec [hunk 2]
```diff
@@ -269,9 +269,31 @@ GRUB_PXE_MODULE_SOURCE=%{buildroot}%{_datadir}/grub2-efi/grubaa64-noprefix.efi
...
+if [ -f /etc/default/grub ] && [ ! -s /etc/default/grub ]; then
+cat << EOF > /etc/default/grub
+# File generated by postscript of grub2 rpm
+GRUB_TIMEOUT=$(grep -w 'set timeout' /boot/grub2/grub.cfg | awk -F'=' 'NR==1{print $NF}')
+GRUB_DISABLE_SUBMENU=true
+GRUB_TERMINAL="serial console"
+GRUB_SERIAL_COMMAND="serial"
+GRUB_CMDLINE_LINUX="rd.auto=1 $(grep mariner_cmdline /boot/mariner.cfg | cut -d= -f2-) lockdown=integrity sysctl.kernel.unprivileged_bpf_disabled=1 $(grep -o "console=[^[:space:]]*" /proc/cmdline |tr '\n' ' ') $(grep systemd_cmdline /boot/systemd.cfg| cut -d= -f2-)"
```
Comment (ddstreet) [inline SPECS/grub2/grub2.spec:287]: definitely should not generate our config based on /boot/mariner.cfg file in running instances; we should actually move our config from static files into grub config and use grub-mkconfig for real to generate grub config in new images

File: SPECS/grub2/grub2.spec [hunk 3]
```diff
@@ -269,9 +269,31 @@ GRUB_PXE_MODULE_SOURCE=%{buildroot}%{_datadir}/grub2-efi/grubaa64-noprefix.efi
...
+GRUB_DISABLE_RECOVERY="true"
+GRUB_DEFAULT="saved"
+GRUB_DISTRIBUTOR="CBL-Mariner"
+EOF
+fi
+
 %postun -p /sbin/ldconfig
 
+%verify (not md5 mtime) /etc/default/grub
```
Comment (ddstreet) [inline SPECS/grub2/grub2.spec:296]: drop this if config file moved to normal %files section

**Review by ddstreet — CHANGES_REQUESTED** [review]
I think this needs more widespread changes to remove our hardcoded grub config in mariner kernel, systemd, etc rpms

File: SPECS/grub2/grub2.spec (hunk 2 above)
Comment (AZaugg) [reply SPECS/grub2/grub2.spec:287]: I can see that there are several ways that kernel comandlines get set,
- via installkernel
- kernel install post scripts via /boot/mariner.cfg
- during initial image creation and template expansion 
- and grub env, laid out by the installer for systemd.cfg

As grub tools were not installed into the base image, I was under the impression that the intent was to keep grub management the way it is. And with this PR I intended to support the current grub management but add compatibility for grub tools for those that wanted to use it moving forward.

Did you want to push towards using grub tools by default?

**Pull request closed without merging**